### PR TITLE
[IR] Add type_check for UnaryOpExpression

### PR DESCRIPTION
**Pull request by strongoier**
Related issue = #3301

Now for
```python
import taichi as ti

ti.init()

@ti.kernel
def func():
    a = 1
    b = ti.ceil(a)

func()
```

The error message is:
```
[Taichi] version 0.8.4, llvm 10.0.0, commit d577b876, linux, python 3.8.10
[Taichi] Starting on arch=x64
[Taichi] Compilation failed
  File "type_test.py", line 8, in func
    b = ti.ceil(a)
TaichiTypeError: 'ceil' takes real inputs only, however 'i32' is provided
```

### CODE REVIEW

File: tests/cpp/ir/frontend_type_inference_test.cpp [hunk 1]
```diff
@@ -47,5 +47,17 @@ TEST(FrontendTypeInference, BinaryOp) {
   EXPECT_EQ(truediv_f64->ret_type, PrimitiveType::f64);
 }
 
+TEST(FrontendTypeInference, UnaryOp) {
```
Comment (k-ye) [inline tests/cpp/ir/frontend_type_inference_test.cpp:50]: FYI there's an `EXPECT_THROW` checker :-) https://google.github.io/googletest/reference/assertions.html#EXPECT_THROW

**Review by k-ye — APPROVED** [review]
LGTM!

File: tests/cpp/ir/frontend_type_inference_test.cpp (hunk 1 above)
Comment (strongoier) [reply tests/cpp/ir/frontend_type_inference_test.cpp:50]: Wow great! I'm just thinking about whether it is better to put the error checks in Python tests or C++ tests. Putting them in Python tests can directly give some illegal examples, while putting them in C++ tests can make `type_check` tests more uniform. WDYT?

Comment (ailzhang) [reply tests/cpp/ir/frontend_type_inference_test.cpp:50]: Haha this is a hard one, I'm more used to putting fine grained tests in C++ and keep python tests sparse but representable. (sorry I know this is vague...

**Review by ailzhang — APPROVED** [review]
Looks great!

File: tests/cpp/ir/frontend_type_inference_test.cpp (hunk 1 above)
Comment (strongoier) [reply tests/cpp/ir/frontend_type_inference_test.cpp:50]: > putting fine grained tests in C++ and keep python tests sparse but representable

Sounds good. Will follow this in future PRs.

**Pull request merged**